src/test_utils/snapshot.rs

```diff
@@ -0,0 +1,279 @@
+use std::collections::HashMap;
+
+use crate::{participants::Participant, protocol::MessageData};
+
+/// Constructor for specific (sender, messages) pairs transmitted during a protocol run
+#[derive(Debug, PartialEq, Clone)]
+struct ReceivedMessageSnapshot {
+    from: Participant,
+    message: MessageData,
+}
+
+impl ReceivedMessageSnapshot {
+    fn new(from: Participant, message: MessageData) -> Self {
+        Self { from, message }
+    }
+}
+
+/// Registers a particular participant's view of the received messages
+struct ParticipantSnapshot {
+    snaps: Vec<ReceivedMessageSnapshot>,
+    read_index: usize,
+}
+
+impl ParticipantSnapshot {
+    pub fn new_empty() -> Self {
+        Self {
+            snaps: Vec::new(),
+            read_index: 0,
+        }
+    }
+
+    fn push_received_message_snapshot(&mut self, snap: ReceivedMessageSnapshot) {
+        self.snaps.push(snap);
+    }
+
+    fn push_message(&mut self, from: Participant, message: MessageData) {
+        let snap = ReceivedMessageSnapshot::new(from, message);
+        self.push_received_message_snapshot(snap);
+    }
+
+    fn read_next_message(&mut self) -> Option<(Participant, MessageData)> {
+        if self.read_index >= self.snaps.len() {
+            return None;
+        }
+        let message_snap = &self.snaps[self.read_index];
+        self.read_index += 1;
+        Some((message_snap.from, message_snap.message.clone()))
+    }
+
+    fn refresh_read(&mut self) {
+        self.read_index = 0;
+    }
+}
+
+/// Used to store the snapshot of all the messages sent during
+/// the communication rounds of a certain protocol
+pub struct ProtocolSnapshot {
+    snapshots: HashMap<Participant, ParticipantSnapshot>,
+}
+
+impl ProtocolSnapshot {
+    /// Creates an empty snapshot
+    pub fn new_empty(participants: Vec<Participant>) -> Self {
+        let snapshots = participants
+            .into_iter()
+            .map(|p| (p, ParticipantSnapshot::new_empty()))
+            .collect::<HashMap<_, _>>();
+        Self { snapshots }
+    }
+
+    /// Adds a messages sent by a sender and to a receiver to the protocol snapshot
+    pub fn push_message(
+        &mut self,
+        to: Participant,
+        from: Participant,
+        message: MessageData,
+    ) -> Option<()> {
+        self.snapshots
+            .get_mut(&to)
+            .map(|snapshot| snapshot.push_message(from, message))
+    }
+
+    /// Reads the next message stored in the snapshot of a particular participant given as input
+    pub fn read_next_snapshot(
```

Comment:
```suggestion
    pub fn read_next_snapshot_for_participant(
```

Reply:
Allow me to also add:

```suggestion
    pub fn read_next_message_for_participant(
```